### PR TITLE
Catalog of things

### PR DESCRIPTION
**Pull request by miltonHenschel**
- [x] Implement the following methods:

  - add_item method (in all classes that have that method)
  - can_be_archived?() in the Item class
  - move_to_archive() in the Item class
  - can_be_archived?() in the Book class
  - can_be_archived?() in the MusicAlbum class
  - can_be_archived?() in the Movie class
  - can_be_archived?() in the Game class

- [x] Add unit tests for all implemented methods.

- [x] Create a main.rb file that will serve as your console app entry-point.

  - Present the user with a list of options to perform.
  - Let users choose an option.
  - If needed, ask for parameters for the option.
  - Have a way to quit the app.

- [x] Implement the following methods in the UI:

  - List all books
  - List all music albums
  - List all movies
  - List of games
  - List all genres
  - List all labels 
  - List all authors 
  - List all sources 
  - Add a book
  - Add a music album
  - Add a movie
  - Add a game

- [x] Preserved data by saving collections in .json files.

- [x] Create a schema.sql file with tables representing the classes in your app.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,177 @@
...
+- Ruby
+- Rspec
+
+
+ ## App ScreenShot
+  ![](Comming Soon)
+ ## Video Preview
+
+ [Video](Comming Soon)
+
```
Comment (ShadyShawkat) [inline README.md:47]: - [ ] Kindly make sure to record a video presentation of the project. The video has to abide by the [following rules](https://github.com/microverseinc/curriculum-ruby/blob/main/group-capstone/catalog_of_my_things.md#:~:text=Once%20you%20have,with%20your%20peers.) given in the requirements.

File: Classes/item.rb [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+  def move_to_archive
+    @archived = can_be_archived?
+  end
+
+  private
+
+  def can_be_archived?
+    Date.today - Date.parse(@publish_date) > 10 * 365
+  end
+end
```
Comment (ShadyShawkat) [inline Classes/item.rb:32]: - [ ] Kindly add a `label` much like how the `author` and `genre` is added. As per [the given UML diagram](https://github.com/microverseinc/curriculum-ruby/blob/main/group-capstone/images/catalog_of_my_things.png).

File: Classes/source.rb [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+    @items << item
+    item.book_source = self
+  end
+
+  def as_hash
+    {
+      'book_source' => @book_source
+    }
+  end
+end
```
Comment (ShadyShawkat) [inline Classes/source.rb:21]: - [ ] As per [this given requirement](https://github.com/microverseinc/curriculum-ruby/blob/main/group-capstone/catalog_of_my_things_divided.md#:~:text=Team%20member%20%234%20(if%20your%20team%20has%203%20members%20%2D%20skip%20those%20requirements)), you're to skip these requirements if the team has less than 4 members. Kindly delete this class.

File: main.rb [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+      puts 'Incorrect! Must be between 1-12. Please try again...'
+    else
+      method_tocall = app.method(method_name)
+      method_tocall.call
+    end
+    menu
+  end
+end
+
+main
```
Comment (ShadyShawkat) [inline main.rb:59]: - [ ] Please make sure that the app works properly. Currently, when I start the app, I see the following options being displayed:

image


- However, when I try to select let's say number `7` to `Add a book`, nothing happens. 
- Please make sure to display [these options](https://github.com/microverseinc/curriculum-ruby/blob/main/group-capstone/catalog_of_my_things.md#:~:text=The%20following%20options,Add%20a%20game) and make sure that each option does its corresponding action.

File: sql/database.sql [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+	last_name VARCHAR(100),
+	items TEXT [],
+	PRIMARY KEY (id)
+);
+
+CREATE TABLE game (
+  	multiplayer BOOLEAN,
+  	last_played_at DATE,
+  	item_id INTEGER REFERENCES item(id)
+) INHERITS (item);
```
Comment (ShadyShawkat) [inline sql/database.sql:39]: - [ ] Please be sure to implement all classes (apart from the `source` and `movie` if your team is less than 4 members) from the given UML diagram which means that the following classes are missing their tables in the schema file:
   - Genre
   - Label
   - MusicAlbum
   - Book

And please also be sure to create indexes for each of the created foreign keys for each table.

File: sql/database.sql [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+CREATE INDEX idx_label_id ON item (label_id)
+CREATE INDEX idx_source_id ON item (source_id)
+
+-- Author and Game classes
+
+CREATE TABLE author (
+	id INT GENERATED BY DEFAULT AS IDENTITY,
+	first_name VARCHAR(100),
+	last_name VARCHAR(100),
+	items TEXT [],
```
Comment (ShadyShawkat) [inline sql/database.sql:31]: - [ ] Please note that you can't translate every detail from the UML diagram into the schema file, as implementing arrays for the `items` here makes sense in the UML diagram and in code but not in the schema file. Please fix this.

File: sql/database.sql (hunk 5 above)
Comment (ShadyShawkat) [inline sql/database.sql:39]: - [ ] Please note that you can't translate every detail from the UML diagram into the schema file, as implementing inheritance here makes sense in the UML diagram and in code but [not in the schema file](https://github.com/microverseinc/curriculum-ruby/blob/main/group-capstone/catalog_of_my_things.md#:~:text=As%20you%20cannot%20implement%20inheritance%20in%20the%20database%20tables). Please fix this.

- [ ] Kindly create an index for the created foreign key in this table.

File: spec/author_spec.rb [hunk 7]
```diff
@@ -0,0 +1,31 @@
+require_relative '../Classes/author'
```
Comment (ShadyShawkat) [inline spec/author_spec.rb:1]: - [ ] Kindly make sure to include tests for all of the created methods in all of the created models and not just the `author` and `game` models, as per [this](https://github.com/microverseinc/curriculum-ruby/blob/main/group-capstone/catalog_of_my_things.md#:~:text=Add%20unit%20tests%20for%20all%20implemented%20methods.) requirement.

**Review by ShadyShawkat — CHANGES_REQUESTED** [review]
## Hi Team, 

Good job so far! 👍🏻 
There are some issues that you still need to work on to go to the next project but you are almost there!

## To Highlight
- All linters are passing ✔️ 
- Correct use of GitFlow ✔️ 
- Professional looking pull request ✔️ 

Image

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: Classes/item.rb (hunk 2 above)
Comment (mrkamin) [reply Classes/item.rb:32]: @ShadyShawkat 
thanks for your good and valuable feedback, as you mention the Label is added.

File: Classes/source.rb (hunk 3 above)
Comment (mrkamin) [reply Classes/source.rb:21]: @ShadyShawkat,
thanks for your good and valuable feedback, as you mention the source class removed.

File: main.rb (hunk 4 above)
Comment (mrkamin) [reply main.rb:59]: @ShadyShawkat 
thanks for your good and valuable feedback, as you mention the bug solved.

File: sql/database.sql (hunk 5 above)
Comment (mrkamin) [reply sql/database.sql]: @ShadyShawkat 
thanks for your good and valuable feedback, the following tables added in the schema file

File: sql/database.sql (hunk 6 above)
Comment (mrkamin) [reply sql/database.sql]: @ShadyShawkat
thanks for your good and valuable feedback, done

File: sql/database.sql (hunk 5 above)
Comment (mrkamin) [reply sql/database.sql]: @ShadyShawkat
thanks for your good and valuable feedback, the index is are created

File: spec/author_spec.rb (hunk 7 above)
Comment (mrkamin) [reply spec/author_spec.rb]: @ShadyShawkat
thanks for your good and valuable feedback, tests are added

**Review by zainsadaqat — APPROVED** [review]
# Pull Request Approved ✅ 🎊 🎉 
## Hello Team! 👋 

## Things to Appreciate 
- [x] Correct Flow ✅ 
- [x] Clean Code 👍 
- [x] No Linter Errors 🔥 
- [x] Descriptive Pull Request and Summary 👌 

### Your project is complete!  There is nothing else to say other than it's time to merge it.

# Congratulations!!! 🎉 🥳 👏 🙌 🎊 ☑️ 



Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
@zainsadaqat 
thanks for your good and valuable feedback

**Pull request merged**